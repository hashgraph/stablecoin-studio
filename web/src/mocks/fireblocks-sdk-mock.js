const FireblocksSdkMock = jest.fn().mockImplementation(() => ({
}));

module.exports = { FireblocksSDK: FireblocksSdkMock };