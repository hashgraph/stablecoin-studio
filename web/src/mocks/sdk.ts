import PublicKey from 'hedera-stable-coin-sdk/build/esm/src/domain/context/account/PublicKey.js';
import ContractId from 'hedera-stable-coin-sdk/build/esm/src/domain/context/contract/ContractId.js';
import BigDecimal from 'hedera-stable-coin-sdk/build/esm/src/domain/context/shared/BigDecimal.js';
import { HederaId } from 'hedera-stable-coin-sdk/build/esm/src/domain/context/shared/HederaId.js';

export const mockedStableCoinsList = {
	coins:[
		{ symbol: 'HBAR', id: '0.0.123' },
		{ symbol: 'EXP', id: '0.0.13234' },
		{ symbol: 'MCC', id: '0.0.48451290'},
 		{ symbol: 'MCC2', id: '0.0.48471242'},
 		{ symbol: 'MCC3', id: '0.0.48478857'}
	]
};

export const mockedSelectedStableCoin = {
	tokenId: HederaId.from('0.0.49319786'),
	name: 'TEST_ACCELERATOR_SC',
	symbol: 'TEST',
	decimals: 6,
	initialSupply: BigDecimal.fromString('1000.', 6),
	totalSupply: BigDecimal.fromString('1000.', 6),
	maxSupply: BigDecimal.fromString('1000', 6),
	proxyAddress: ContractId.from('0.0.49319785'),
	evmProxyAddress: '0000000000000000000000000000000002f08f69',
	treasury: HederaId.from('0.0.49319785'),
	paused: false,
	deleted: false,
	freezeDefault: false,
	autoRenewAccount: HederaId.from('0.0.49071854'),
	autoRenewAccountPeriod: 90,
	adminKey: ContractId.from('0.0.49319785'),
	kycKey: undefined,
	freezeKey: ContractId.from('0.0.49319785'),
	wipeKey: ContractId.from('0.0.49319785'),
	supplyKey: ContractId.from('0.0.49319785'),
	pauseKey: ContractId.from('0.0.49319785'),
};
export const mockedStableCoinCapabilities = {
	coin: mockedSelectedStableCoin,
	capabilities: [
		{ operation: 'Rescue', access: 1 },
		{ operation: 'Cash_in', access: 1 },
		{ operation: 'Burn', access: 1 },
		{ operation: 'Wipe', access: 1 },
		{ operation: 'Pause', access: 1 },
		{ operation: 'Unpause', access: 1 },
		{ operation: 'Freeze', access: 1 },
		{ operation: 'Unfreeze', access: 1 },
		{ operation: 'Delete', access: 1 },
		{ operation: 'Role_Management', access: 1 },
		{ operation: 'Admin Role', access: 1 },
		{ operation: 'Admin Role', access: 1 },
	],
	account: {
		id: HederaId.from('0.0.48471385' ),
		publicKey: new PublicKey({
			key: '03cce92867cd5e08c67da3fdd0bbae217f7ac73237e2d9058f7551e4ed6d9bf5ce',
			type: 'ECDSA_SECP256K1',
		}),
	},
};

export const mockedWalletData = {
  topic: 'da96975a-ebd9-4382-98f8-41fe25bc2b6b',
  pairingString: 'eWNvbiI6ImNzbHRpQWNjb3VudCI6ZmFsc2V9',
  encryptionKey: 'XdXXecXX-XcXc-XbaX-bXXd-XXXbXXXXcXXX',
  savedPairings: [
    {
      accountIds: ['0.0.48450590'],
      metadata: {
        name: 'HashPack',
        description: 'An example wallet',
        icon: 'https://www.hashpack.app/img/logo.svg',
        url: 'chrome&#45;extension&#58;&#47;&#47;gjagmgiddbbciopjhllkdnddhcglnemk',
        encryptionKey: 'XdXXecXX-XcXc-XbaX-bXXd-XXXbXXXXcXXX',
      },
      network: 'testnet',
      topic: 'da96975a-ebd9-4382-98f8-41fe25bc2b6b',
      origin: 'chrome-extension://gjagmgiddbbciopjhllkdnddhcglnemk',
      encryptionKey: 'XdXXecXX-XcXc-XbaX-bXXd-XXXbXXXXcXXX',
      lastUsed: 1664888808450,
    },
  ],
}

export const mockedFoundWallets = ['HashPack'];