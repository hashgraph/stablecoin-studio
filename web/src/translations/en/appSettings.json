{
	"title": "Configure custom RPC/MIRROR",
	"subtitle": "Choose an option",
	"name": "Name",
	"rpc": "RPC",
	"mirrorNode": "Mirror node",
	"url": "Url",
	"apiKey": "Api Key",
	"addMirror": "Add new service",
	"addRPC": "Add RPC",
	"removeMirror": "Remove Mirror Node",
	"removeRpc": "Remove RPC",
	"header": "Header",
	"mirrorNameExists": "Mirror name already exists, if you want to update it remove it first and recreate it",
	"rpcNameExists": "RPC name already exists, if you want to update it remove it first and recreate it"
}
