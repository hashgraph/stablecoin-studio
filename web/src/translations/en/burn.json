{
    "title": "Burn tokens operation",
    "operationTitle": "Operation details",
    "amountLabel": "Enter the amount",
    "amountPlaceholder": "0",
    "originAccountPlaceholder": "0.0.0",
    "modalSuccessDesc": "{{amount}} tokens have been burned.",
    "modalAction": {
        "subtitle": "Operation details review",
        "originAccount": "Account",
        "amount": "Amount",
        "accept": "Burn tokens"
    }
}