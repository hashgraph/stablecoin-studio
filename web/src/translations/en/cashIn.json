{
    "title": "Cash in operation",
    "operationTitle": "Operation details",
    "amountLabel": "Introduce amount",
    "amountPlaceholder": "0",
    "destinationAccountLabel": "Destination account",
    "destinationAccountPlaceholder": "0.0.0",
    "modalSuccessDesc": "The cash in operation finished successfully.",
    "modalAction": {
        "subtitle": "Operation details review",
        "destinationAccount": "Destination account",
        "amount": "Amount",
        "accept": "Cash in tokens"
    }
}