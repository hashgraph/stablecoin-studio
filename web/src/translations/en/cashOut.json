{
    "title": "Cash out operation",
    "operationTitle": "Operation details",
    "amountLabel": "How much tokens?",
    "amountPlaceholder": "Introduce amount",
    "originAccountLabel": "From which account tokens will be withdrawn?",
    "originAccountPlaceholder": "Introduce account",
    "modalSuccessDesc": "The cash out operation finished successfully.",
    "modalAction": {
        "subtitle": "Operation details review",
        "originAccount": "Account to cash out",
        "amount": "Amount",
        "accept": "Make cash out"
    }
}