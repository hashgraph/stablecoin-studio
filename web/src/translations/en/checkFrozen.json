{
	"title": "Check account frozen status",
	"operationTitle": "Operation details",
	"accountLabel": "Enter the account",
	"accountPlaceholder": "0.0.0",
	"modalIsFrozen": "The account {{ account }} is frozen for the token.",
	"modalIsNotFrozen": "The account {{ account }} is unfrozen for the token.",
	"modalAction": {
		"subtitle": "Operation details review",
		"account": "Account",
		"accept": "Get account frozen status"
	}
}
