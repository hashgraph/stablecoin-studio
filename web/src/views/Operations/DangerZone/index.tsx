import { Box, Heading, Stack, HStack, Text } from '@chakra-ui/react';
import { useTranslation } from 'react-i18next';
import BaseContainer from '../../../components/BaseContainer';
import GridDirectAction from '../../../components/GridDirectAction';
import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
	SELECTED_TOKEN_ROLES,
	SELECTED_WALLET_CAPABILITIES,
	SELECTED_WALLET_COIN,
	SELECTED_WALLET_PAIRED_ACCOUNTID,
	walletActions,
} from '../../../store/slices/walletSlice';
import type { DirectActionProps } from '../../../components/DirectAction';
import type { StableCoinCapabilities } from '@hashgraph/stablecoin-npm-sdk';
import {
	PauseRequest,
	DeleteRequest,
	Operation,
	Access,
	StableCoinRole,
} from '@hashgraph/stablecoin-npm-sdk';
import type { IAccountToken } from '../../../interfaces/IAccountToken';
import type { IExternalToken } from '../../../interfaces/IExternalToken';
import type { ModalsHandlerActionsProps } from '../../../components/ModalsHandler';
import SDKService from '../../../services/SDKService';
import { useNavigate } from 'react-router-dom';
import { NamedRoutes } from '../../../Router/NamedRoutes';
import { RouterManager } from '../../../Router/RouterManager';

const DangerZoneOperations = () => {
	const { t } = useTranslation('operations');

	const navigate = useNavigate();
	const dispatch = useDispatch();

	const selectedStableCoin = useSelector(SELECTED_WALLET_COIN);
	const accountId = useSelector(SELECTED_WALLET_PAIRED_ACCOUNTID);
	const capabilities: StableCoinCapabilities | undefined = useSelector(
		SELECTED_WALLET_CAPABILITIES,
	);
	const roles = useSelector(SELECTED_TOKEN_ROLES)!;

	const [disabledFeatures, setDisabledFeatures] = useState({
		pause: false,
		delete: false,
		unpause: false,
	});
	const [errorPauseOperation, setErrorPauseOperation] = useState('');
	const [errorUnpauseOperation, setErrorUnpauseOperation] = useState('');
	const [errorDeleteOperation, setErrorDeleteOperation] = useState('');
	const [errorTransactionUrl, setErrorTransactionUrl] = useState();

	const [requestPause] = useState(
		new PauseRequest({
			tokenId: selectedStableCoin?.tokenId?.toString() ?? '',
		}),
	);
	const [requestDelete] = useState(
		new DeleteRequest({
			tokenId: selectedStableCoin?.tokenId?.toString() ?? '',
		}),
	);

	const [isPaused, setPaused] = useState(false);
	const [isDeleted, setDeleted] = useState(false);

	useEffect(() => {
		if (selectedStableCoin) {
			checkTokenStatus();
			getAvailableFeatures();
		}
	}, [selectedStableCoin]);

	const checkTokenStatus = async () => {
		setPaused(selectedStableCoin?.paused || false);
		setDeleted(selectedStableCoin?.deleted || false);
	};

	const getAvailableFeatures = () => {
		let isExternalToken = false;
		const tokensAccount = localStorage?.tokensAccount;
		if (tokensAccount) {
			const tokensAccountParsed = JSON.parse(tokensAccount);
			if (tokensAccountParsed) {
				const myAccount = tokensAccountParsed.find(
					(acc: IAccountToken) => acc.id === accountId?.toString(),
				);
				if (myAccount) {
					const externalToken = myAccount?.externalTokens.find(
						(coin: IExternalToken) => coin.id === selectedStableCoin?.tokenId?.toString(),
					);
					if (externalToken) {
						isExternalToken = true;
					}
				}
			}
		}

		const operations = capabilities?.capabilities.map((x) => x.operation);
		const canPause = operations?.includes(Operation.PAUSE);
		const areDisabled: {
			pause: boolean;
			unpause: boolean;
			delete: boolean;
		} = {
			pause: !isExternalToken
				? (!canPause || (canPause && selectedStableCoin?.paused)) ?? false
				: (!operations?.includes(Operation.PAUSE) ||
						(operations?.includes(Operation.PAUSE) &&
							getAccessByOperation(Operation.PAUSE) !== Access.HTS &&
							!roles.includes(StableCoinRole.PAUSE_ROLE)) ||
						selectedStableCoin?.paused) ??
				  false,
			unpause: !isExternalToken
				? (!canPause || (canPause && !selectedStableCoin?.paused)) ?? false
				: (!operations?.includes(Operation.PAUSE) ||
						(operations?.includes(Operation.PAUSE) &&
							getAccessByOperation(Operation.PAUSE) !== Access.HTS &&
							!roles.includes(StableCoinRole.PAUSE_ROLE)) ||
						!selectedStableCoin?.paused) ??
				  false,
			delete: !isExternalToken
				? (!canPause || selectedStableCoin?.paused || selectedStableCoin?.deleted) ?? false
				: (!operations?.includes(Operation.DELETE) ||
						(operations?.includes(Operation.DELETE) &&
							getAccessByOperation(Operation.DELETE) !== Access.HTS &&
							!roles.includes(StableCoinRole.DELETE_ROLE)) ||
						selectedStableCoin?.paused ||
						selectedStableCoin?.deleted) ??
				  false,
		};

		setDisabledFeatures(areDisabled);

		function getAccessByOperation(operation: Operation): Access | undefined {
			return (
				capabilities?.capabilities.filter((capability) => {
					return capability.operation === operation;
				})[0].access ?? undefined
			);
		}
	};

	const handlePause: ModalsHandlerActionsProps['onConfirm'] = async ({
		onSuccess,
		onError,
		onLoading,
	}) => {
		try {
			onLoading();
			if (!selectedStableCoin?.proxyAddress || !selectedStableCoin?.tokenId) {
				onError();
				return;
			}

			await SDKService.pause(requestPause);
			dispatch(walletActions.setPausedToken(true));
			dispatch(walletActions.setSelectedStableCoin({ ...selectedStableCoin, paused: true }));
			onSuccess();
		} catch (error: any) {
			setErrorTransactionUrl(error.transactionUrl);
			setErrorPauseOperation(error.message);
			onError();
		}
	};

	const handleUnpause: ModalsHandlerActionsProps['onConfirm'] = async ({
		onSuccess,
		onError,
		onLoading,
	}) => {
		try {
			onLoading();
			if (!selectedStableCoin?.proxyAddress || !selectedStableCoin?.tokenId) {
				onError();
				return;
			}

			await SDKService.unpause(requestPause);
			dispatch(walletActions.setPausedToken(false));
			dispatch(walletActions.setSelectedStableCoin({ ...selectedStableCoin, paused: false }));
			onSuccess();
		} catch (error: any) {
			setErrorTransactionUrl(error.transactionUrl);
			setErrorUnpauseOperation(error.message);
			onError();
		}
	};

	const handleDelete: ModalsHandlerActionsProps['onConfirm'] = async ({
		onSuccess,
		onError,
		onLoading,
	}) => {
		try {
			onLoading();
			if (!selectedStableCoin?.proxyAddress || !selectedStableCoin?.tokenId) {
				onError();
				return;
			}

			await SDKService.delete(requestDelete);
			onSuccess();
			dispatch(walletActions.setDeletedToken(true));
			dispatch(walletActions.setSelectedStableCoin({ ...selectedStableCoin, deleted: true }));
			RouterManager.to(navigate, NamedRoutes.Operations);
		} catch (error: any) {
			setErrorTransactionUrl(error.transactionUrl);
			setErrorDeleteOperation(error.message);
			onError();
		}
	};

	const directActions: DirectActionProps[] = [
		{
			icon: 'Pause',
			handleAction: handlePause,
			title: t('dangerZone.pauseOperation'),
			isDisabled: disabledFeatures?.pause,
			errorNotification: errorPauseOperation,
			errorTransactionUrl,
			operationTranslate: 'pause',
		},
		{
			icon: 'Play',
			handleAction: handleUnpause,
			title: t('dangerZone.unpauseOperation'),
			isDisabled: disabledFeatures?.unpause,
			errorNotification: errorUnpauseOperation,
			errorTransactionUrl,
			operationTranslate: 'unpause',
		},
		{
			icon: 'Trash',
			handleAction: handleDelete,
			title: t('dangerZone.deleteOperation'),
			isDisabled: disabledFeatures?.delete,
			errorNotification: errorDeleteOperation,
			errorTransactionUrl,
			operationTranslate: 'delete',
		},
	];

	const filteredDirectActions = directActions.filter((access) => !access.isDisabled);

	return (
		<Stack h='full'>
			<HStack spacing={6} w='full'>
				{isPaused && (
					<Text
						fontSize='16px'
						color='brand.secondary'
						fontWeight={700}
						align='right'
						w='full'
						as='i'
						data-testid='paused-subtitle'
					>
						{t('pausedToken')}
					</Text>
				)}
				{isDeleted && (
					<Text
						fontSize='16px'
						color='brand.secondary'
						fontWeight={700}
						align='right'
						w='full'
						as='i'
						data-testid='deleted-subtitle'
					>
						{t('deletedToken')}
					</Text>
				)}
			</HStack>
			<BaseContainer title={t('title')}>
				<Box p={{ base: 4, md: '128px' }}>
					<Heading fontSize='20px' fontWeight='600' mb={14} data-testid='subtitle'>
						{t('subtitle')}
					</Heading>
					<GridDirectAction directActions={filteredDirectActions} />
				</Box>
			</BaseContainer>
		</Stack>
	);
};

export default DangerZoneOperations;
