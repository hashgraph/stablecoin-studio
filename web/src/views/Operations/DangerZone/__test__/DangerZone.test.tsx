import { render } from '../../../../test/index';
import en from '../../../../translations/en/operations.json';
import DangerZone from '..';
import configureMockStore from 'redux-mock-store';
import { mockedStableCoinCapabilities } from '../../../../mocks/sdk';
import userEvent from '@testing-library/user-event';
import { StableCoinRole } from '@hashgraph/stablecoin-npm-sdk';

const translations = en;

describe(`<${DangerZone.name} />`, () => {
	test('should render correctly', () => {
		const component = render(<DangerZone />);

		expect(component.asFragment()).toMatchSnapshot();
	});

	test('should render titles', () => {
		const component = render(<DangerZone />);

		expect(component.getByTestId('subtitle')).toHaveTextContent(translations.subtitle);
	});

	test('should have options', async () => {
		const mockStore = configureMockStore();
		const selectedStableCoin = {
			initialSupply: 10,
			tokenId: '0.0.12345',
			proxyAddress: '0.0.12345',
			reserveAmount: '10',
			paused: false,
		};

		const store = mockStore({
			wallet: {
				selectedStableCoin,
				capabilities: mockedStableCoinCapabilities,
				data: {
					account: {
						id: 'id',
					},
				},
				roles: [StableCoinRole.PAUSE_ROLE, StableCoinRole.DELETE_ROLE],
			},
		});
		localStorage.setItem(
			'tokensAccount',
			JSON.stringify([
				{
					id: 'id',
					externalTokens: [
						{
							id: '0.0.12345',
							symbol: 'symbol',
						},
					],
				},
			]),
		);

		const component = render(<DangerZone />, store);

		const pause = component.getAllByTestId('direct-access-Pause');
		await userEvent.click(pause[0]);

		const confirmModalButton = component.getByTestId('modal-action-confirm-button');
		await userEvent.click(confirmModalButton);

		const deleteButton = component.getAllByTestId('direct-access-Delete');
		await userEvent.click(deleteButton[0]);

		const confirmButtons = component.getAllByTestId('modal-action-confirm-button');
		await userEvent.click(confirmButtons[1]);
	}, 60000);
});
