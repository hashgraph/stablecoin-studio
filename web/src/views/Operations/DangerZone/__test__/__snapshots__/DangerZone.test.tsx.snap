// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DangerZoneOperations /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-16oka5y"
  >
    <div
      class="chakra-stack css-5vgp34"
    />
    <div
      class="chakra-stack css-16oka5y"
    >
      <h2
        class="chakra-heading css-bqsusc"
        data-testid="base-container-heading"
      >
        Operations
      </h2>
      <div
        class="css-1b6lda7"
        data-testid="base-container-children"
      >
        <div
          class="css-src872"
        >
          <h2
            class="chakra-heading css-qltjox"
            data-testid="subtitle"
          >
            Choose an operation
          </h2>
          <div
            class="css-nxcc5s"
          />
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
